services/collaboration/pkg/connector/fileconnector.go

```diff
@@ -472,6 +525,482 @@ func (f *FileConnector) UnLock(ctx context.Context, lockID string) (string, erro
+	// assume the target is a full name
+	wopiContext, err := middleware.WopiContextFromCtx(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	logger := zerolog.Ctx(ctx).With().
+		Str("PutTarget", target).
+		Logger()
+
+	// stat the current file in order to get the reference of the parent folder
+	oldStatRes, err := f.gwc.Stat(ctx, &providerv1beta1.StatRequest{
+		Ref: wopiContext.FileReference,
+	})
+	if err != nil {
+		logger.Error().Err(err).Msg("PutRelativeFileRelative: stat failed")
+		return nil, nil, err
+	}
+
+	if oldStatRes.GetStatus().GetCode() != rpcv1beta1.Code_CODE_OK {
+		logger.Error().
+			Str("StatusCode", oldStatRes.GetStatus().GetCode().String()).
+			Str("StatusMsg", oldStatRes.GetStatus().GetMessage()).
+			Msg("PutRelativeFileRelative: stat failed with unexpected status")
+		return nil, nil, NewConnectorError(500, oldStatRes.GetStatus().GetCode().String()+" "+oldStatRes.GetStatus().GetMessage())
+	}
+
+	targetPath := utils.MakeRelativePath(target)
+	// need to change the file reference of the wopicontext to point to the new path
+	wopiContext.FileReference = &providerv1beta1.Reference{
+		ResourceId: oldStatRes.GetInfo().GetParentId(),
+		Path:       targetPath,
+	}
+
+	// create a new context for the modified wopicontext
+	newLogger := logger.With().Str("NewFileReference", wopiContext.FileReference.String()).Logger()
+	newCtx := middleware.WopiContextToCtx(newLogger.WithContext(ctx), wopiContext)
+
+	var conError *ConnectorError
+	// try to put the file. It mustn't return a 400 or 409
+	lockID, err := ccs.PutFile(newCtx, stream, streamLength, "")
+	if err != nil {
+		// if the error isn't a connectorError, fail the request
+		if !errors.As(err, &conError) {
+			newLogger.Error().Err(err).Msg("PutRelativeFileRelative: put file failed")
+			return nil, nil, err
+		}
+
+		if conError.HttpCodeOut == 409 {
+			if err := f.adjustWopiReference(ctx, &wopiContext, newLogger); err != nil {
+				return nil, nil, err
+			}
+			// if conflict generate a different name and retry.
+			// this should happen only once
+			wopiSrcURL, err2 := f.generateWOPISrc(ctx, wopiContext, newLogger)
+			if err2 != nil {
+				newLogger.Error().
+					Err(err2).
+					Str("LockID", lockID).
+					Msg("PutRelativeFileRelative: error generating the WOPISrc parameter for conflict response")
+				return nil, nil, err
+			}
+
+			actualFilename, _ := f.extractFilenameAndPrefix(target)
+			finalTarget := f.generatePrefix() + " " + actualFilename
+			headers := &PutRelativeHeaders{
+				ValidTarget: finalTarget,
+				LockID:      lockID,
+			}
+			response := &PutRelativeResponse{
+				Name: target,
+				Url:  wopiSrcURL.String(),
+			}
+			newLogger.Error().
+				Err(err).
+				Str("LockID", lockID).
+				Msg("PutRelativeFileRelative: error conflict")
+			return response, headers, err
+		} else {
+			// TODO: code 400 might happen, what to do?
```

Comment:
What about this TODO?

Reply:
I'll remove that comment. The "relative" code was mostly copied from the "suggested" one because the behavior is very similar, if in this case, the "relative" code can return a 400 error code. The error will just be forwarded from the `PutFile` request.